### PR TITLE
DNS: do not query CNAME if A succeeded already

### LINKED ISSUE #48048: [Bug]: DNS CNAME record is always queried, even when the query for an A record already succeeded
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

I am looking at this code:

https://github.com/nextcloud/server/blob/4281ce6fa1bb8235426099d720734d2394bec203/lib/private/Http/Client/DnsPinMiddleware.php#L61-L89

> [!IMPORTANT]
>
> I am **by no means** a PHP expert, please tell me if I am mistaken.

To me, the way the code is constructed would indicate that CNAME is always queried, even when A succeeded. This aligns with what I am observing in my DNS server logs:

```txt
linux/amd64, go1.21.11, a6338e9
[INFO] 127.0.0.1:39942 - 18117 "HINFO IN 5684395984608177492.63579198951733077. udp 55 false 512" NXDOMAIN qr,rd,ra 130 0.054606916s
[INFO] 10.10.0.38:42508 - 60413 "CNAME IN updates.nextcloud.com. udp 50 false 1200" NOERROR qr,rd,ra 116 0.039573751s
[INFO] 10.10.0.38:46316 - 26341 "CNAME IN updates.nextcloud.com.cloud.svc.cluster. udp 68 false 1200" NXDOMAIN qr,aa,rd,ra 132 0.000151319s
[INFO] 10.10.0.38:42074 - 34897 "CNAME IN updates.nextcloud.com.cluster. udp 58 false 1200" NXDOMAIN qr,aa,rd,ra 122 0.000110026s
[INFO] 10.10.0.38:54477 - 22629 "CNAME IN apps.nextcloud.com.cloud.svc.cluster. udp 65 false 1200" NXDOMAIN qr,aa,rd,ra 129 0.000151308s
[INFO] 10.10.0.38:49188 - 36978 "CNAME IN www.nextcloud.com.cloud.svc.cluster. udp 64 false 1200" NXDOMAIN qr,aa,rd,ra 128 0.000266391s
[INFO] 10.10.0.38:48706 - 44842 "CNAME IN www.nextcloud.com.cluster. udp 54 false 1200" NXDOMAIN qr,aa,rd,ra 118 0.000134442s
[INFO] 10.10.0.38:40937 - 47170 "CNAME IN nextcloud.com. udp 42 false 1200" NOERROR qr,rd,ra 108 0.040937649s
[INFO] 10.10.0.38:57557 - 28949 "CNAME IN nextcloud.com.svc.cluster. udp 54 false 1200" NXDOMAIN qr,aa,rd,ra 118 0.000163416s
```

Wouldn't it make more sense to skip CNAME when `canHaveCnameRecord` is `false`, and we should set `canHaveCnameRecord` outside the loop?

### Steps to reproduce

1. Run NC
2. Look at DNS server logs

### Expected behavior

When `canHaveCnameRecord` is `false`, we should not query for CNAME.

### Nextcloud Server version

29

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.3

### Web server

Apache (supported)

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
I will deliver upon request; I do not think the config is important for this issue.
```

### List of activated Apps

```shell
I will deliver upon request; I do not think the config is important for this issue.
```

### Nextcloud Signing status

```shell
I will deliver upon request; I do not think the config is important for this issue.
```

### Nextcloud Logs

```shell
I will deliver upon request; I do not think the config is important for this issue.
```

### Additional info

This is not a showstopper in any way; I just think the behavior is neither expected nor desirable.

### PR DESCRIPTION
**Pull request by georglauterbach**
Fixes: #48048

## Summary

When an A query succeeds, querying for CNAME again is

1. useless at best, and
2. creating unnecessary log entries in the DNS server.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Comment by solracsf** [comment]
Tests are failing:

```
There was 1 failure:

1) lib\Http\Client\DnsPinMiddlewareTest::testIgnoreSubdomainForSoaQuery
Failed asserting that actual size 3 matches expected size 4.

/home/runner/actions-runner/_work/server/server/tests/lib/Http/Client/DnsPinMiddlewareTest.php:540
```

**Comment by georglauterbach** [comment]
> Tests are failing:

I hope that is fixed with [e3d7d84](https://github.com/nextcloud/server/pull/48675/commits/e3d7d846038d667ee9ae3bfb198a655bb11a6836); I am not a PHP programmer, though; hence, I am not sure whether `assertNotContains` is used correctly here.

**Comment by solracsf** [comment]
Another failure:

```
There was 1 failure:

1) lib\Http\Client\DnsPinMiddlewareTest::testIgnoreSubdomainForSoaQuery
Failed asserting that an array does not contain 'subsubdomain.subdomain.example.com134217728'.
```

**Comment by georglauterbach** [comment]
I updated the tests

**Comment by georglauterbach** [comment]
Checks are waiting for approval

File: tests/lib/Http/Client/DnsPinMiddlewareTest.php [hunk 1]
```diff
@@ -537,10 +537,11 @@ static function (RequestInterface $request, array $options) {
 			['nextcloud' => ['allow_local_address' => false]]
 		);
 
-		$this->assertCount(4, $dnsQueries);
+		$this->assertCount(3, $dnsQueries);
 		$this->assertContains('example.com' . DNS_SOA, $dnsQueries);
 		$this->assertContains('subsubdomain.subdomain.example.com' . DNS_A, $dnsQueries);
 		$this->assertContains('subsubdomain.subdomain.example.com' . DNS_AAAA, $dnsQueries);
-		$this->assertContains('subsubdomain.subdomain.example.com' . DNS_CNAME, $dnsQueries);
+		// CNAME should not be queried of A or AAAA succeeded already
```
Comment (artonge) [inline tests/lib/Http/Client/DnsPinMiddlewareTest.php:544]:
```suggestion
		// CNAME should not be queried if A or AAAA succeeded already
```

**Review by artonge — COMMENTED** [review]
Looks got besides the typo in the comment :)

File: tests/lib/Http/Client/DnsPinMiddlewareTest.php (hunk 1 above)
Comment (georglauterbach) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php:544]: 👍🏼 

Should I also rewrite the commit messages? The action that checks on conventional commits fails.

Comment (artonge) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php:544]: Indeed, you can squash the commits and prefix it with `fix: ` to respect the convention: https://www.conventionalcommits.org/en/v1.0.0/

Comment (georglauterbach) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php]: Done

Comment (artonge) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php]: Please remove the merge commit

File: lib/private/Http/Client/DnsPinMiddleware.php [hunk 2]
```diff
@@ -57,17 +57,21 @@ private function dnsResolve(string $target, int $recursionCount) : array {
...
+
 			if ($this->negativeDnsCache->isNegativeCached($target, $dnsType)) {
 				continue;
 			}
 
 			$dnsResponses = $this->dnsGetRecord($target, $dnsType);
-			$canHaveCnameRecord = true;
 			if ($dnsResponses !== false && count($dnsResponses) > 0) {
 				foreach ($dnsResponses as $dnsResponse) {
 					if (isset($dnsResponse['ip'])) {
```
Comment (come-nc) [inline lib/private/Http/Client/DnsPinMiddleware.php:77]: The line 85 below `$canHaveCnameRecord = true;` can be removed no? It’s in an if branch that checks the boolean is already true.
I know this part was not touched by the PR but it helps understanding the code to remove this useless line.

Comment (georglauterbach) [reply lib/private/Http/Client/DnsPinMiddleware.php:77]: Indeed; I will remove the line

File: tests/lib/Http/Client/DnsPinMiddlewareTest.php (hunk 1 above)
Comment (georglauterbach) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php]: > Please remove the merge commit

I pressed on "Update Branch" here on GitHub - I guess this should not be done?

Comment (georglauterbach) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php]: I removed the merge commits again; the commits are only mine now

**Review by artonge — APPROVED** [review]

File: tests/lib/Http/Client/DnsPinMiddlewareTest.php (hunk 1 above)
Comment (artonge) [reply tests/lib/Http/Client/DnsPinMiddlewareTest.php]: > I pressed on "Update Branch" here on GitHub - I guess this should not be done?

You can rebase your branch if you want. There is the option in the dropdown menu next to the "Update branch" button.

**Review by come-nc — APPROVED** [review]

**Comment by come-nc** [comment]
@georglauterbach Thanks a lot! This problem has been bothering people for a long time, I hope this will help reduce the chatter in the logs.

**Pull request merged**